File: javascript/astar.js
/**
 * Algoritmo A* (A-estrela) para caminho mínimo em grades 2D.
 *
 * Espelho exato da implementação Python (python/astar.py): mesma ordem de
 * exploração de vizinhos, mesmo critério de desempate no heap
 * (f, depois h, depois ordem de inserção) e, portanto, exatamente os mesmos
 * nós expandidos — o que torna a comparação de tempos entre as duas
 * linguagens uma comparação justa do MESMO trabalho algorítmico.
 *
 * Movimento em 4 direções com custo unitário; heurística de Manhattan
 * (admissível e consistente), o que garante a otimalidade do caminho.
 */

// Ordem fixa: cima, baixo, esquerda, direita (idêntica ao Python).
export const DIRECOES = [[-1, 0], [1, 0], [0, -1], [0, 1]];

export const LIVRE = 0;
export const OBSTACULO = 1;

/** Distância de Manhattan entre duas células [linha, coluna]. */
export function manhattan(a, b) {
  return Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]);
}

/**
 * Heap binário de mínimo com desempate lexicográfico (f, h, ordem).
 * Cada elemento é [f, h, ordem, indiceDaCelula].
 */
export class MinHeap {
  constructor() {
    this.itens = [];
  }

  get tamanho() {
    return this.itens.length;
  }

  static _menor(a, b) {
    if (a[0] !== b[0]) return a[0] < b[0];
    if (a[1] !== b[1]) return a[1] < b[1];
    return a[2] < b[2];
  }

  push(item) {
    const itens = this.itens;
    itens.push(item);
    let i = itens.length - 1;
    while (i > 0) {
      const paiIdx = (i - 1) >> 1;
      if (MinHeap._menor(itens[i], itens[paiIdx])) {
        [itens[i], itens[paiIdx]] = [itens[paiIdx], itens[i]];
        i = paiIdx;
      } else {
        break;
      }
    }
  }

  pop() {
    const itens = this.itens;
    const topo = itens[0];
    const ultimo = itens.pop();
    if (itens.length > 0) {
      itens[0] = ultimo;
      let i = 0;
      for (;;) {
        const esq = 2 * i + 1;
        const dir = 2 * i + 2;
        let menor = i;
        if (esq < itens.length && MinHeap._menor(itens[esq], itens[menor])) menor = esq;
        if (dir < itens.length && MinHeap._menor(itens[dir], itens[menor])) menor = dir;
        if (menor === i) break;
        [itens[i], itens[menor]] = [itens[menor], itens[i]];
        i = menor;
      }
    }
    return topo;
  }
}

/**
 * Executa o A* na grade.
 *
 * @param {number[][]} grid  matriz (0 = livre, 1 = obstáculo)
 * @param {[number, number]} start  célula de origem [linha, coluna]
 * @param {[number, number]} goal   célula de destino [linha, coluna]
 * @returns {{caminho: Array<[number,number]>|null, custo: number, expandidos: number}}
 *   caminho da origem ao destino (null se não existe), custo em passos
 *   (Infinity se não existe) e número de nós expandidos.
 */
export function astar(grid, start, goal) {
  const linhas = grid.length;
  const colunas = grid[0].length;

  for (const [nome, [r, c]] of [["origem", start], ["destino", goal]]) {
    if (!(r >= 0 && r < linhas && c >= 0 && c < colunas)) {
      throw new RangeError(`${nome} (${r}, ${c}) fora da grade ${linhas}x${colunas}`);
    }
  }

  if (grid[start[0]][start[1]] === OBSTACULO || grid[goal[0]][goal[1]] === OBSTACULO) {
    return { caminho: null, custo: Infinity, expandidos: 0 };
  }

  // Células indexadas por r * colunas + c. Map/Set espelham os dicionários
  // da versão Python: a memória cresce só com a região explorada, mantendo a
  // mesma assintótica nas duas implementações (sem inicialização Theta(n)).
  const idxOrigem = start[0] * colunas + start[1];
  const idxDestino = goal[0] * colunas + goal[1];

  const gScore = new Map([[idxOrigem, 0]]);
  const pai = new Map();
  const expandido = new Set();

  const heap = new MinHeap();
  let ordem = 0;
  const h0 = manhattan(start, goal);
  heap.push([h0, h0, ordem++, idxOrigem]);

  let expandidos = 0;

  while (heap.tamanho > 0) {
    const [, , , v] = heap.pop();
    if (expandido.has(v)) continue; // entrada obsoleta no heap
    expandido.add(v);
    expandidos++;

    if (v === idxDestino) {
      return { caminho: reconstruir(pai, v, colunas), custo: gScore.get(v), expandidos };
    }

    const r = Math.floor(v / colunas);
    const c = v % colunas;
    const gv = gScore.get(v);

    for (const [dr, dc] of DIRECOES) {
      const nr = r + dr;
      const nc = c + dc;
      if (nr >= 0 && nr < linhas && nc >= 0 && nc < colunas && grid[nr][nc] === LIVRE) {
        const u = nr * colunas + nc;
        if (expandido.has(u)) continue;
        const novoG = gv + 1;
        const gAtual = gScore.get(u);
        if (gAtual === undefined || novoG < gAtual) {
          gScore.set(u, novoG);
          pai.set(u, v);
          const hu = Math.abs(nr - goal[0]) + Math.abs(nc - goal[1]);
          heap.push([novoG + hu, hu, ordem++, u]);
        }
      }
    }
  }

  return { caminho: null, custo: Infinity, expandidos };
}

/** Reconstrói o caminho seguindo os pais, do destino para a origem. */
function reconstruir(pai, v, colunas) {
  const caminho = [];
  while (v !== undefined) {
    caminho.push([Math.floor(v / colunas), v % colunas]);
    v = pai.get(v);
  }
  caminho.reverse();
  return caminho;
}

File: python/astar.py
"""Algoritmo A* (A-estrela) para caminho mínimo em grades 2D.

Problema resolvido: encontrar o caminho de menor custo entre dois pontos de uma
grade com obstáculos, movendo-se em 4 direções (cima, baixo, esquerda, direita)
com custo unitário por passo.

O A* é uma busca de melhor-primeiro que ordena a fronteira pela função
f(v) = g(v) + h(v), onde g(v) é o custo real acumulado da origem até v e h(v)
é uma estimativa heurística do custo de v até o destino. Com heurística
admissível (nunca superestima) e consistente (h(u) <= custo(u,v) + h(v)),
o A* é ótimo: o primeiro caminho encontrado é o de menor custo.

Heurística usada: distância de Manhattan — admissível e consistente para
movimento em 4 direções com custo unitário.

Pseudocódigo:
    A*(grade, origem, destino):
        aberta <- heap-mínimo ordenado por (f, h, ordem de inserção)
        g[origem] <- 0
        insere origem na aberta com f = h(origem)
        enquanto aberta não vazia:
            v <- remove-mínimo(aberta)
            se v já foi expandido: continua          # entrada obsoleta no heap
            marca v como expandido
            se v == destino: retorna caminho reconstruído
            para cada vizinho u de v (4 direções, livre e dentro da grade):
                g' <- g[v] + 1
                se g' < g[u]:
                    g[u] <- g'; pai[u] <- v
                    insere u na aberta com f = g' + h(u)
        retorna "sem caminho"

Complexidade (V células livres, E arestas; em grades E <= 4V, logo E = O(V)):
    - Pior caso:   O(V log V) tempo (cada aresta pode gerar uma inserção no
                   heap de tamanho O(V)), O(V) memória.
    - Melhor caso: Omega(L) expansões, onde L é o comprimento do caminho ótimo
                   (heurística perfeita guia direto ao destino).
"""

from __future__ import annotations

import heapq
from itertools import count

# Ordem fixa de exploração dos vizinhos: cima, baixo, esquerda, direita.
# A MESMA ordem é usada na implementação JavaScript para que as duas
# implementações expandam exatamente os mesmos nós, na mesma ordem.
DIRECOES = ((-1, 0), (1, 0), (0, -1), (0, 1))

LIVRE = 0
OBSTACULO = 1


def manhattan(a: tuple[int, int], b: tuple[int, int]) -> int:
    """Distância de Manhattan entre duas células (linha, coluna)."""
    return abs(a[0] - b[0]) + abs(a[1] - b[1])


def astar(grid: list[list[int]],
          start: tuple[int, int],
          goal: tuple[int, int]):
    """Executa o A* na grade.

    Args:
        grid: matriz de inteiros (0 = livre, 1 = obstáculo).
        start: célula de origem (linha, coluna).
        goal: célula de destino (linha, coluna).

    Returns:
        (caminho, custo, expandidos):
        - caminho: lista de células da origem ao destino, ou None se não existe.
        - custo: número de passos do caminho (inf se não existe).
        - expandidos: quantidade de nós retirados do heap e processados
          (métrica de trabalho do algoritmo, idêntica entre as linguagens).
    """
    linhas, colunas = len(grid), len(grid[0])

    for nome, (r, c) in (("origem", start), ("destino", goal)):
        if not (0 <= r < linhas and 0 <= c < colunas):
            raise ValueError(f"{nome} {(r, c)} fora da grade {linhas}x{colunas}")

    if grid[start[0]][start[1]] == OBSTACULO or grid[goal[0]][goal[1]] == OBSTACULO:
        return None, float("inf"), 0

    # Desempate do heap: menor f, depois menor h (favorece nós mais próximos
    # do destino), depois ordem de inserção (determinismo total — essencial
    # para a comparação justa entre Python e JavaScript).
    ordem = count()
    h0 = manhattan(start, goal)
    aberta: list[tuple[int, int, int, tuple[int, int]]] = [(h0, h0, next(ordem), start)]
    g_score: dict[tuple[int, int], int] = {start: 0}
    pai: dict[tuple[int, int], tuple[int, int]] = {}
    expandido: set[tuple[int, int]] = set()
    expandidos = 0

    while aberta:
        _f, _h, _, v = heapq.heappop(aberta)
        if v in expandido:
            continue  # entrada obsoleta (o nó já saiu com g menor)
        expandido.add(v)
        expandidos += 1

        if v == goal:
            return _reconstruir(pai, v), g_score[v], expandidos

        gv = g_score[v]
        r, c = v
        for dr, dc in DIRECOES:
            nr, nc = r + dr, c + dc
            if 0 <= nr < linhas and 0 <= nc < colunas and grid[nr][nc] == LIVRE:
                u = (nr, nc)
                if u in expandido:
                    continue
                novo_g = gv + 1
                if novo_g < g_score.get(u, float("inf")):
                    g_score[u] = novo_g
                    pai[u] = v
                    hu = manhattan(u, goal)
                    heapq.heappush(aberta, (novo_g + hu, hu, next(ordem), u))

    return None, float("inf"), expandidos


def _reconstruir(pai: dict, v: tuple[int, int]) -> list[tuple[int, int]]:
    """Reconstrói o caminho do destino até a origem seguindo os pais."""
    caminho = [v]
    while v in pai:
        v = pai[v]
        caminho.append(v)
    caminho.reverse()
    return caminho
